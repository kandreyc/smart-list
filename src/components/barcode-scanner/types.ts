
export interface BarcodeScanResult {
    code: string | null;
    format: string;
}
