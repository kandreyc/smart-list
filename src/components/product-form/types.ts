type ProductFormData = {
    barcode: string;
    barcodeType: string;
    generalName: string;

    fullName: string | null;
    country: string | null;
    companyName: string | null;
};

export default ProductFormData;
