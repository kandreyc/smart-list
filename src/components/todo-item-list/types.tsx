import PriceData from "../../entity/PriceData";
import Product from "../../entity/Product";

export class ProductPriceData {
    latestPrice: number;
    perCounterpartyPrice: {[id: string]: PriceData};

    constructor () {
        this.latestPrice = 0;
        this.perCounterpartyPrice = {};
    }

    setCounterpartyPrice(counterparty: string, priceData: PriceData) {
        this.perCounterpartyPrice[counterparty] = priceData;
    }
}

export class Store {
    readonly id: number;
    readonly name: string;

    constructor(id: number, name: string) {
        this.id = id;
        this.name = name;
    }
}

class TodoItem {
    id: number;
    generalName: string;
    quantity: number;
    targetProduct?: Product;
    isBought: boolean;
    priceData: ProductPriceData;

    constructor (id: number, generalName: string) {
        this.id = id;
        this.generalName = generalName;
        this.quantity = 1;
        this.isBought = false;
        this.priceData = new ProductPriceData();
    }

    static fromProduct(product: Product, quantity?: number): TodoItem {
        let todoItem = new TodoItem(Date.now(), product.productGeneralName);

        todoItem.quantity = quantity? quantity : 1;
        todoItem.targetProduct = product;
        todoItem.isBought = false;
        return todoItem;
    }

    static from(json: any): TodoItem {
        let todoItem = new TodoItem(json.id, json.generalName);

        todoItem.quantity = json.quantity;
        todoItem.targetProduct = Product.from(json.targetProduct);
        todoItem.isBought = json.isBought;
        return todoItem;
    }
}

export default TodoItem;
