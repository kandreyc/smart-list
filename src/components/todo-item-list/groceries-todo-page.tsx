import React, { useState } from "react";
import { TodoItemListContextProvider } from "../../pages/groceries-todo/context/TodoItemListContextProvider";
import useGroceriesTodoController, { Mode } from "./components/use-groceries-todo-controller";
import GroceriesTodoPlanningModeView from "./components/groceries-todo-planning";
import GroceriesTodoPurchasingModeView from "./components/groceries-todo-purchasing";
import GroceriesTodoToolbar from "./components/groceries-todo-tool-bar";
import { GroceriesTodoStoreContextProvider } from "./components/groceries-todo-store-context";


const GroceriesTodoPage = () => {

    const controller = useGroceriesTodoController();

    const switchMode = (isPurchaseMode: boolean) => {
        if (isPurchaseMode) {
            controller.enablePurchasingMode();
        } else {
            controller.enablePlanningMode();
        }
    }

    return (<>
        <TodoItemListContextProvider>
            <GroceriesTodoStoreContextProvider>
                <GroceriesTodoToolbar onPurchaseModeToggle={switchMode} />
                {controller.currentMode === Mode.PLANNING && <>
                    <GroceriesTodoPlanningModeView />
                </>}

                {controller.currentMode === Mode.PURCHASING && <>
                    <GroceriesTodoPurchasingModeView />
                </>}
            </GroceriesTodoStoreContextProvider>
        </TodoItemListContextProvider>
    </>);
}


export default GroceriesTodoPage;
