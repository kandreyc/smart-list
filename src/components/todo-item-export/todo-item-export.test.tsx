import { shallow } from "enzyme";
import Category from "../../entity/category";
import { Store } from "../todo-item-list/types";
import ExportItem from "./export-item";
import TodoItemExport, { getExportResult } from "./todo-item-export";

const groceriesCategory = { id: 0, name: "Groceries" };
const snacksCategory = { id: 1, name: "Snacks" };

const auchanStore = { id: 0, name: "Auchan" } as Store;
const zabkaStore = { id: 1, name: "Zabka" } as Store;

function createExportItem(
    generalName: string,
    purchasedPrice?: number,
    category?: Category,
    purchasedStore?: Store
) {
    return new ExportItem(
        1,
        generalName,
        1,
        true,
        null,
        purchasedPrice || null,
        purchasedStore || null,
        category || null,
        null
    );
}

describe("TodoItemExport", () => {
    const todoItemExport = () => <TodoItemExport />;

    it("should render", () => {
        const wrapper = shallow(todoItemExport());
        expect(wrapper).toMatchSnapshot();
    });

    describe("functions", () => {
        it("should group by category", () => {
            const milk = createExportItem("Milk", 3.44, groceriesCategory);
            const bread = createExportItem("Bread", 2.55, groceriesCategory);
            const pringles = createExportItem("Pringles", 3.55, snacksCategory);
            const exportItems = [milk, bread, pringles];

            const result = getExportResult(exportItems);

            expect(result).toEqual([{
                category: groceriesCategory.name,
                storeName: "undefined",
                exportItemsByStore: [milk, bread],
            }, {
                category: snacksCategory.name,
                storeName: "undefined",
                exportItemsByStore: [pringles],
            }]);
        });

        it("should group by store", () => {
            const milk = createExportItem("Milk", 3.44, undefined, auchanStore);
            const bread = createExportItem("Bread", 2.55, undefined, auchanStore);
            const pringles = createExportItem("Pringles", 3.55, undefined, zabkaStore);
            const exportItems = [milk, bread, pringles];

            const result = getExportResult(exportItems);

            expect(result).toEqual([{
                category: "undefined",
                storeName: auchanStore.name,
                exportItemsByStore: [milk, bread],
            }, {
                category: "undefined",
                storeName: zabkaStore.name,
                exportItemsByStore: [pringles],
            }]);
        });

        it("should group by category and store", () => {
            const milk = createExportItem("Milk", 3.44, groceriesCategory, auchanStore);
            const bread = createExportItem("Bread", 2.55, groceriesCategory, auchanStore);
            const lays = createExportItem("Lays", 2.55, snacksCategory, auchanStore);
            const pringles = createExportItem("Pringles", 3.55, snacksCategory, zabkaStore);
            const exportItems = [milk, bread, pringles, lays];

            const result = getExportResult(exportItems);

            expect(result).toHaveLength(3);

            expect(result).toContainEqual({
                category: groceriesCategory.name,
                storeName: auchanStore.name,
                exportItemsByStore: [milk, bread],
            });

            expect(result).toContainEqual({
                category: snacksCategory.name,
                storeName: auchanStore.name,
                exportItemsByStore: [lays],
            });

            expect(result).toContainEqual({
                category: snacksCategory.name,
                storeName: zabkaStore.name,
                exportItemsByStore: [pringles],
            });
        });
    });
});
