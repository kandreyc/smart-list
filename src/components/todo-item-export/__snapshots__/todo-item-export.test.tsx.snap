// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItemExport should render 1`] = `
<Fragment>
  <ExportItemList
    exportItems={Array []}
  />
</Fragment>
`;
